### DISCUSSION TITLE
Allow arrow keys to relocate search bar cursor

### DISCUSSION THREAD

**Discussion by aaron-bradshaw** [ppy/osu#30978] [Ideas]
Currently in lazer the search bars in settings and beatmap selection cannot be navigated with the arrow keys. This makes it very annoying to fix small errors like the one shown below because you have to erase up to the point of the mistake and retype the rest or manually move the cursor with your mouse.
![image](https://github.com/user-attachments/assets/60087322-a58f-4021-ba01-d6b969bb7b04)
I don't know if this is intended behavior to allow for left and right to scroll the beatmap list, but the up and down arrows already provide the exact same functionality while also being more visually intuitive because the beatmap list is already vertical. This also does not apply to the settings search box because you cannot interact with the beatmap selection while the settings panel is open. The arrow keys can be used to change slider values, but you cannot have a slider and the text box focused at the same time.

This isn't really the biggest deal in the world but it is really annoying when you try and use the arrow keys in the settings menu and nothing happens. It's one of the small pain points that, if fixed, could make even stubborn stable players switch to lazer because it "feels better'.

**Comment by N0Y0U2114** [comment]
I think it should be a modifier as for using keyboard only you use the arrow keys (L/R for going between mapsets U/D for going between maps). 
Currently Shift + Arrows selects and Ctrl + Arrows goes to Start or end, so maybe **Alt + Arrow** to move cursor within text?

**Comment by tarkhodvai** [comment]
i know that this discussion thread is old and i might accidentally anger some people who think that posting on old threads like this is bad, but i really just want to give two cents about how the current beatmap song select navigation system is clunky to use, especially when you want to search for something.

considering that it has been 6 months since the last comment was published, the osu!lazer client have been getting better over the days, with song select v2 being the most important update as of to date (8 June 2025), and in each of the update i have been getting of since i started playing seriously back in march, the updates are very good and honestly, i really do enjoy the update. however, one roadblock that throws the song select screen, even to this very day, off from "perfection" is that the searchbox for the song select screen is very frustrating to use unless if you use a mouse to directly tell the searchbox cursor where it should go.

yes, as mentioned above, there are ways to scroll the search cursor without using your mouse to directly put the cursor on the place the user want, such as ctrl + left/right arrow keys scrolls text per-word, and shift + left/right arrow keys scrolls _**and**_ selects text per-letter (note the "and" being emphasized), but the lack of a key combination, or ideally a key by itself, for per-letter text scrolling in a searchbox is a bit of a pet peeve to me.

as suggested by this https://github.com/ppy/osu/discussions/30978#discussioncomment-11491923 by N0Y0U2114, assigning alt + left/right arrow keys to scroll text per-letter is a pretty good suggestion. as for my suggestions for this issue, i basically have two of them in mind.

the first approach to this issue is to make the searchbox clickable, and when you click the searchbox, your left and right arrow keys became the navigation key for the search cursor. this preserves the current beatmap carousel navigation system by basically highlighting whether or not the searchbox is clicked or not. consistency might be an issue here, but this behavior is standard in many UI interfaces, and this is basically the most plausible option for this issue, as this only requires changes to the searchbox element to be clickable, and an indicator on whether or not the searchbox is clicked or not. 

the second approach to this issue, requires more effort to implement it. considering that in song select v2, pressing up/down arrow key now only moves the cursor, and you need to press enter to select it, however, left/right arrow key still selects the item per beatmapset. to make the left/right arrow key useable for searchbox navigation, the previous function of the keys need to go somewhere else, preferably within the up/down arrow keys. with this in mind, i figured out several keybinds that can be used for beatmap carousel navigation to free up the left/right key for searchbox navigation usage, here as follows:

- up/down arrow key moves the cursor per item, shift + up/down arrow keys moves the cursor per beatmapset (consistent in regards to behavior, but removes the ability to be able to automatically select the item since ctrl + up/down and alt + up/down are taken by HT/DT mod speed adjustment and volume control respectively)
- up/down arrow key moves the cursor per item, shift + up/down arrow keys selects per beatmapset (inconsistent behavior between two, thus making this unlikely)
- up/down arrow key selects per item, shift + up/down arrow keys selects per beatmapset (reverts the behavior back to song select v1 version, unlikely since it clashes with the recent update's intention)
- up/down arrow key moves the cursor per item, shift + up/down arrow keys selects per item (removes the ability to being able to select song by beatmapset, unlikely since this removes a feature that people might still use)

with the issue sorted out, now we can assign the left/right arrow keys to searchbox navigation system. this also have a benefit of consolidating both beatmap carousel navigation and searchbox navigation under one set of keys each, with up and down arrows being for beatmap carousel and left and right arrows being for seachbox navigation respectively.

**Comment by frenzibyte** [comment]
The settings search bar should not be used for writing more than two/three short keywords maximum. If you messed up by typo'ing, consider pressing Ctrl+A, backspace, then write again. There's no need for mouse. There is no point for adding seeking keys in search bars.

For song select search bar, I would honestly say the same still applies. Write enough that you can select the beatmap by few arrow key presses, then go select it.

**Reply by tarkhodvai** [reply]
so to fix a typo, you are basically suggesting that i should write the keywords all over again? this is even more effort than doing shift + arrow keys and then ctrl + the same arrow key as previous to seek to the place where you make a typo.
